File: go/BinaryTreePreOrder/BinaryTreePreOrder.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time -> 0(N) Space -> 0(N)
func preorderTraversal(root *TreeNode) []int {
	array := []int{}
	preOrder(root, &array)
	return array
}

func preOrder(root *TreeNode, array *[]int) {
	if root == nil {
		return
	}

	*array = append(*array, root.Val)
	preOrder(root.Left, array)
	preOrder(root.Right, array)
}

func preorderTraversalIterative(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	stack := []*TreeNode{root}
	array := []int{}

	for len(stack) != 0 {
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		array = append(array, root.Val)
		if root.Right != nil {
			stack = append(stack, root.Right)
		}

		if root.Left != nil {
			stack = append(stack, root.Left)
		}
	}
	return array
}

func main() {
	node2 := &TreeNode{Val: 3}
	node1 := &TreeNode{Val: 2}
	root := &TreeNode{Val: 1, Left: node1, Right: node2}
	fmt.Println(preorderTraversal(root))
	fmt.Println(preorderTraversalIterative(root))
}

File: python/binaryTreePreorder.py
class TreeNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(N) Space -> 0(N)
def preorderTraversal(root) -> list[int]:
    array = []
    preorder(root, array)
    return array


def preorder(root, array):
    if root is None:
        return

    array.append(root.val)
    preorder(root.left, array)
    preorder(root.right, array)


def preorderIterative(root) -> list[int]:
    if root is None:
        return []
    stack = [root]
    array = []
    while len(stack) != 0:
        root = stack.pop()
        array.append(root.val)
        if root.right is not None:
            stack.append(root.right)
        if root.left is not None:
            stack.append(root.left)
    return array
